{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "CATAStropy.ipynb",
      "private_outputs": true,
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true,
      "authorship_tag": "ABX9TyOYfEuqia0g8HbtihKMMgkb",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "gpuClass": "standard"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/ccdmb/catastrophy/blob/master/CATAStropy.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# CATAStrophy\n",
        "\n",
        "This is a simple GUI for the CATAStrophy pipeline.\n",
        "\n",
        "- Paper: https://doi.org/10.3389/fmicb.2019.03088\n",
        "- GitHub: https://github.com/ccdmb/catastrophy\n",
        "\n",
        "To run the CATAStrophy pipeline, click `Runtime>Restart and run all` in the colab menu. Then:\n",
        "\n",
        "1. In the file browser to the left of the screen, right click on the `fastas` folder and select upload.\n",
        "2. Upload any fasta files that you want to run through CATAStrophy.\n",
        "3. Click the `refresh` button below.\n",
        "4. If you uploaded any undesired files or you wish to re-run the program with only a subset of the fastas, you can select/unselect the files using the select menu (control-click to select/deselect individuals, shift-click to select many files).\n",
        "5. Select the version of dbCAN to search, we have trained different models for different versions so the results may be slightly different.\n",
        "6. Click \"Run\" to run the CATAStrophy pipeline.\n",
        "\n",
        "The program will download the correct version of dbCAN, run HMMER, and run CATAStrophy for you. The program typically takes roughly 5 minutes per-proteome per CPU. Coogle colab notebooks have a default of 2 CPUs so you can expect 1 or 2 proteomes to run in about 5 minutes, or 3 (or 4) proteomes to run in a bit over 10 minutes etc.\n",
        "\n",
        "If you wish, you can add re-run steps 1-6 to add new proteomes. The notebook will cache already computed results for you.\n",
        "\n",
        "**IMPORTANT: do not re-run any code cells, as it may result in odd behaviour. Please only use the buttons etc provided. If you do re-run code, please select `Runtime>Restart and run all` as before. You can hit the Run button and your previous results should still be there.**\n",
        "\n",
        "We include a basic interactive table and scatterplot to browse your results.\n",
        "To change the axes or nomenclatures in the plot, select the changes you want to make and click the `render` button. Your new proteomes are labelled with the \"nomenclature\" `user`. You can highlight specific nomenclatures by clicking the circles in the legend (Shift-click to highlight multiple). Double clicking any circle in the legend will return the default plot. \n",
        "\n",
        "You can download the results by clicking the `Download` button, or by selecting the files you want from the file browser on the left.\n",
        "Note that the download button compresses the results into a zip file, and can take some time to download. Scroll the the bottom of the output to see a small progress bar. Occasionally it won't work and you may have to download from the file browser at the left yourself (tip: right click in the browser and select `refresh`).\n",
        "\n",
        "**Note: Google colab environments will automatically delete the \"Runtime\" (i.e. all of the saved files and installed software) after a few hours of non-use. Please make sure your download your results when you are finished.**\n",
        "\n",
        "Troubleshooting:\n",
        "\n",
        "- If you get an error about \"Runtimes\" or the notebook doesn't appear to be doing anything, select the dropdown menu to the top right of the page and select `connect to runtime`."
      ],
      "metadata": {
        "id": "Ar5B2W9RRptq"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title\n",
        "import os\n",
        "import shutil\n",
        "from os.path import join as pjoin\n",
        "from subprocess import run\n",
        "\n",
        "\n",
        "def install_hmmer():\n",
        "    hmmscan_absent = shutil.which(\"hmmscan\") is None\n",
        "    hmmpress_absent = shutil.which(\"hmmpress\") is None\n",
        "\n",
        "    if hmmscan_absent or hmmpress_absent:\n",
        "        try:\n",
        "            t = run(\n",
        "                \"apt-get install hmmer\",\n",
        "                shell=True,\n",
        "                check=True,\n",
        "                capture_output=True\n",
        "            )\n",
        "        except Exception as e:\n",
        "            print(t.stdout.decode())\n",
        "            print(t.stderr.decode())\n",
        "            raise e\n",
        "\n",
        "    return\n",
        "\n",
        "\n",
        "def install_catastrophy():\n",
        "    try:\n",
        "        import catas  # noqa: W0611\n",
        "    except ModuleNotFoundError:\n",
        "        try:\n",
        "            t = run(\n",
        "                \"pip install catastrophy\",\n",
        "                shell=True,\n",
        "                check=True,\n",
        "                capture_output=True\n",
        "            )\n",
        "        except Exception as e:\n",
        "            print(t.stdout.decode())\n",
        "            print(t.stderr.decode())\n",
        "            raise e\n",
        "    return\n",
        "\n",
        "\n",
        "install_hmmer()\n",
        "install_catastrophy()\n",
        "\n",
        "\n",
        "from catas.pipeline.main import runner as pipeline  # noqa: E402\n",
        "from catas.data import Version, nomenclatures  # noqa: E402\n",
        "from google.colab import files  # noqa: E402\n",
        "from google.colab import data_table  # noqa: E402\n",
        "data_table.enable_dataframe_formatter()\n",
        "\n",
        "import ipywidgets as widgets  # noqa: E402\n",
        "from IPython.display import display  # noqa: E402\n",
        "\n",
        "import pandas as pd  # noqa: E402\n",
        "import numpy as np  # noqa: E402\n",
        "import altair as alt  # noqa: E402\n",
        "\n",
        "os.makedirs(\"fastas\", exist_ok=True)\n",
        "fastas = os.listdir(\"fastas\")\n",
        "\n",
        "refresh_button = widgets.Button(description=\"Refresh\", disabled=False)\n",
        "fasta_select = widgets.SelectMultiple(\n",
        "    options=fastas,\n",
        "    value=tuple(fastas),\n",
        "    description=\"Select files\"\n",
        ")\n",
        "\n",
        "version = widgets.Dropdown(\n",
        "    options=Version,\n",
        "    value=Version.latest(),\n",
        "    description=\"Version:\"\n",
        ")\n",
        "\n",
        "run_button = widgets.Button(description=\"Run\", disabled=len(fastas) < 1)\n",
        "run_output = widgets.Output()\n",
        "\n",
        "download_button = widgets.Button(description=\"Download\", disabled=True)\n",
        "\n",
        "nomenclature_select = widgets.Select(\n",
        "    options=[\"nomenclature1\", \"nomenclature2\", \"nomenclature3\"],\n",
        "    value=\"nomenclature3\",\n",
        "    description=\"Select nomenclature\"\n",
        ")\n",
        "\n",
        "xaxis_select = widgets.Select(\n",
        "    options=[f\"pc{p:0>2}\" for p in range(1, 17)],\n",
        "    value=\"pc01\",\n",
        "    description=\"Select X-axis\"\n",
        ")\n",
        "\n",
        "yaxis_select = widgets.Select(\n",
        "    options=[f\"pc{p:0>2}\" for p in range(1, 17)],\n",
        "    value=\"pc02\",\n",
        "    description=\"Select Y-axis\"\n",
        ")\n",
        "\n",
        "render_button = widgets.Button(description=\"Render\", disabled=False)\n",
        "\n",
        "table_output = widgets.Output()\n",
        "plot_output = widgets.Output()\n",
        "\n",
        "\n",
        "# Not used, It wasn't reliable.\n",
        "\"\"\"\n",
        "def upload_fastas():\n",
        "    old_dir = os.getcwd()\n",
        "    os.chdir(pjoin(old_dir, \"fastas\"))\n",
        "    uploaded = files.upload()\n",
        "    os.chdir(old_dir)\n",
        "    return\n",
        "\"\"\"\n",
        "\n",
        "\n",
        "def on_refresh(change):\n",
        "    os.makedirs(\"fastas\", exist_ok=True)\n",
        "\n",
        "    old_values = fasta_select.get_interact_value()\n",
        "    old_options = fasta_select.options\n",
        "    to_exclude = set(old_options).difference(old_values)\n",
        "\n",
        "    fasta_select.options = os.listdir(\"fastas\")\n",
        "    fasta_select.value = old_values + tuple(\n",
        "        k for k\n",
        "        in os.listdir(\"fastas\")\n",
        "        if k not in to_exclude\n",
        "    )\n",
        "    return\n",
        "\n",
        "\n",
        "def on_select_value_change(change):\n",
        "    run_button.disabled = len(fasta_select.get_interact_value()) < 1\n",
        "\n",
        "\n",
        "def get_results_df():\n",
        "    pca = pd.read_csv(\"results/pca.tsv\", sep=\"\\t\", na_values=\".\")\n",
        "    columns = list(pca.columns)\n",
        "    columns[0] = \"label\"\n",
        "    pca.columns = columns\n",
        "\n",
        "    new_indices = pca.loc[pca.genome.isnull(), ].index.values\n",
        "    pca.loc[\n",
        "        new_indices,\n",
        "        [\"nomenclature1\", \"nomenclature2\", \"nomenclature3\"]\n",
        "    ] = \"user\"\n",
        "    del columns\n",
        "    return pca\n",
        "\n",
        "\n",
        "def display_results_df(pca):\n",
        "    new_indices = pca.loc[pca.genome.isnull(), ].index.values\n",
        "    training_indices = np.setdiff1d(pca.index.values, new_indices)\n",
        "\n",
        "    with table_output:\n",
        "        display(data_table.DataTable(\n",
        "            pca.loc[\n",
        "                np.concatenate([new_indices, training_indices]),\n",
        "                [c for c in pca.columns if not c.startswith(\"pc\")]\n",
        "            ],\n",
        "            include_index=False,\n",
        "            num_rows_per_page=10,\n",
        "        ))\n",
        "\n",
        "\n",
        "def on_render(b):\n",
        "    plot_it()\n",
        "    return\n",
        "\n",
        "\n",
        "def plot_it():\n",
        "    nomenclature = nomenclature_select.value\n",
        "    selection = alt.selection_multi(fields=[nomenclature])\n",
        "    x = xaxis_select.value\n",
        "    y = yaxis_select.value\n",
        "\n",
        "    color = alt.condition(\n",
        "        selection,\n",
        "        alt.Color(f'{nomenclature}:N', legend=None),\n",
        "        alt.value('lightgray')\n",
        "    )\n",
        "\n",
        "    scatter = alt.Chart(pca).mark_circle(size=50).encode(\n",
        "        x=f\"{x}:Q\",\n",
        "        y=f\"{y}:Q\",\n",
        "        color=color,\n",
        "        tooltip=[\"label:N\"]\n",
        "    ).properties(\n",
        "        width=800,\n",
        "        height=500,\n",
        "    ).interactive()\n",
        "\n",
        "    order = [\"user\"] + nomenclatures()[nomenclature]\n",
        "\n",
        "    legend = alt.Chart(pca).mark_circle(size=100).encode(\n",
        "        y=alt.Y(\n",
        "            f'{nomenclature}:O',\n",
        "            axis=alt.Axis(orient='right'),\n",
        "            sort=order\n",
        "        ),\n",
        "        color=color\n",
        "    ).properties(\n",
        "        width=50,\n",
        "        height=200,\n",
        "    ).add_selection(\n",
        "        selection\n",
        "    )\n",
        "    plot_output.clear_output()\n",
        "\n",
        "    with plot_output:\n",
        "        display((scatter | legend).configure_axis(\n",
        "            labelFontSize=14,\n",
        "            titleFontSize=16,\n",
        "        ))\n",
        "    return\n",
        "\n",
        "\n",
        "def on_run(change):\n",
        "    run_output.clear_output()\n",
        "    selected = fasta_select.get_interact_value()\n",
        "    run_fastas = [pjoin(\"fastas\", s) for s in selected]\n",
        "\n",
        "    download_button.disabled = False\n",
        "    global pca\n",
        "    pca = get_results_df()\n",
        "\n",
        "    with run_output:\n",
        "        pipeline(\n",
        "            infiles=run_fastas,\n",
        "            version=version.value,\n",
        "            outdir=\"results\",\n",
        "            hmms=None,\n",
        "            hmmscan_path=shutil.which(\"hmmscan\"),\n",
        "            hmmpress_path=shutil.which(\"hmmpress\"),\n",
        "            ncpu=-1,\n",
        "            correct=False,\n",
        "            quiet=False,\n",
        "        )\n",
        "\n",
        "        render_button.on_click(on_render)\n",
        "        selectors = widgets.HBox([\n",
        "            nomenclature_select,\n",
        "            xaxis_select,\n",
        "            yaxis_select\n",
        "        ])\n",
        "        layout = widgets.VBox([\n",
        "            table_output,\n",
        "            selectors,\n",
        "            render_button,\n",
        "            plot_output,\n",
        "        ])\n",
        "        display(layout)\n",
        "        display_results_df(pca)\n",
        "        plot_it()\n",
        "    return\n",
        "\n",
        "\n",
        "def on_download(change):\n",
        "    import datetime\n",
        "    now = datetime.datetime.now().strftime(\"%Y%m%d-%H%M%S\")\n",
        "    basename = f\"{now}-catastrophy_results\"\n",
        "    shutil.make_archive(basename, \"zip\", \"results\")\n",
        "    print(basename)\n",
        "    files.download(f\"{basename}.zip\")\n",
        "    return\n",
        "\n",
        "\n",
        "refresh_button.on_click(on_refresh)\n",
        "fasta_select.observe(on_select_value_change)\n",
        "run_button.on_click(on_run)\n",
        "download_button.on_click(on_download)\n",
        "\n",
        "buttons = widgets.HBox([refresh_button, run_button, download_button])\n",
        "vbox = widgets.VBox([buttons, fasta_select, version, run_output])\n",
        "display(vbox)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "toO52j3YBDSK"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}